File: advent_of_code/2024/python/02.py
def is_safe(level):
    return all(1 <= abs(a - b) <= 3 for a, b in zip(level, level[1:])) and (
        all(a > b for a, b in zip(level, level[1:]))
        or all(a < b for a, b in zip(level, level[1:]))
    )


def is_almost_safe(level):
    return any(is_safe(level[:i] + level[i + 1 :]) for i in range(len(level)))


def part1(text):
    return sum(is_safe(list(map(int, line.split()))) for line in text.splitlines())


def part2(text):
    return sum(
        is_safe(level) or is_almost_safe(level)
        for line in text.splitlines()
        if (level := list(map(int, line.split())))
    )

File: advent_of_code/2024/rust/src/02.rs
fn is_safe(level: &Vec<i32>) -> bool {
    let all_diff_valid = level
        .windows(2)
        .all(|pair| matches!(pair, &[a, b] if(1..=3).contains(&(a - b).abs())));
    let all_increasing = level
        .windows(2)
        .all(|pair| matches!(pair, &[a, b] if a < b));
    let all_decreasing = level
        .windows(2)
        .all(|pair| matches!(pair, &[a, b] if a > b));
    all_diff_valid && (all_increasing || all_decreasing)
}

fn is_almost_safe(level: &Vec<i32>) -> bool {
    (0..level.len()).any(|i| {
        let mut temp = level.to_vec();
        temp.remove(i);
        is_safe(&temp)
    })
}

fn part1(text: &str) -> usize {
    text.lines()
        .map(|line| {
            let level: Vec<i32> = line
                .split_whitespace()
                .map(|num| num.parse().unwrap())
                .collect();
            is_safe(&level)
        })
        .filter(|&is_safe| is_safe)
        .count()
}

fn part2(text: &str) -> usize {
    text.lines()
        .map(|line| {
            let level: Vec<i32> = line
                .split_whitespace()
                .map(|num| num.parse().unwrap())
                .collect();
            is_safe(&level) || is_almost_safe(&level)
        })
        .filter(|&is_safe| is_safe)
        .count()
}

fn main() {
    aoc2024::solve(part1, part2);
}
